pkg/container/vector/vector.go

```diff
@@ -72,1608 +65,757 @@ func (v *Vector) GetIsBin() bool {
-		fillDefaultValue[int64](v)
+		return extend[int64](v, rows, mp)
 	case types.T_uint8:
-		fillDefaultValue[uint8](v)
+		return extend[uint8](v, rows, mp)
 	case types.T_uint16:
-		fillDefaultValue[uint16](v)
+		return extend[uint16](v, rows, mp)
 	case types.T_uint32:
-		fillDefaultValue[uint32](v)
+		return extend[uint32](v, rows, mp)
 	case types.T_uint64:
-		fillDefaultValue[uint64](v)
+		return extend[uint64](v, rows, mp)
 	case types.T_float32:
-		fillDefaultValue[float32](v)
+		return extend[float32](v, rows, mp)
 	case types.T_float64:
-		fillDefaultValue[float64](v)
+		return extend[float64](v, rows, mp)
+	case types.T_char, types.T_varchar, types.T_json, types.T_blob, types.T_text:
+		return extend[types.Varlena](v, rows, mp)
 	case types.T_date:
-		fillDefaultValue[types.Date](v)
+		return extend[types.Date](v, rows, mp)
 	case types.T_datetime:
-		fillDefaultValue[types.Datetime](v)
+		return extend[types.Datetime](v, rows, mp)
 	case types.T_time:
-		fillDefaultValue[types.Time](v)
+		return extend[types.Time](v, rows, mp)
 	case types.T_timestamp:
-		fillDefaultValue[types.Timestamp](v)
+		return extend[types.Timestamp](v, rows, mp)
 	case types.T_decimal64:
-		fillDefaultValue[types.Decimal64](v)
+		return extend[types.Decimal64](v, rows, mp)
 	case types.T_decimal128:
-		fillDefaultValue[types.Decimal128](v)
+		return extend[types.Decimal128](v, rows, mp)
 	case types.T_uuid:
-		fillDefaultValue[types.Uuid](v)
+		return extend[types.Uuid](v, rows, mp)
 	case types.T_TS:
-		fillDefaultValue[types.TS](v)
+		return extend[types.TS](v, rows, mp)
 	case types.T_Rowid:
-		fillDefaultValue[types.Rowid](v)
-	case types.T_char, types.T_varchar, types.T_json, types.T_blob, types.T_text:
-		fillDefaultValue[types.Varlena](v)
+		return extend[types.Rowid](v, rows, mp)
 	default:
-		panic("unsupported type in FillDefaultValue")
+		panic(fmt.Sprintf("unexpect type %s for function vector.PreExtend", v.typ))
+	}
+}
+
+// Dup use to copy an identical vector
+func (v *Vector) Dup(m *mpool.MPool) (*Vector, error) {
+	var err error
+
+	w := &Vector{
+		typ:    v.typ,
+		isBin:  v.isBin,
+		class:  v.class,
+		length: v.length,
+		nsp:    v.nsp.Clone(),
+		col:    new(reflect.SliceHeader),
+	}
+
+	sz := v.typ.TypeSize()
+	data := (*(*[]byte)(unsafe.Pointer(v.col)))[:v.length/sz]
+	// Copy v.data, note that this should work for Varlena type
+	// as because we will copy area next and offset len will stay
+	// valid for long varlena.
+	if len(data) > 0 {
+		var ndata []byte
+
+		if ndata, err = m.Alloc(int(len(data))); err != nil {
+			return nil, err
+		}
+		copy(ndata, data)
+		*(*[]byte)(unsafe.Pointer(w.col)) = ndata
 	}
+	if len(v.area) > 0 {
+		if w.area, err = m.Alloc(int(len(v.area))); err != nil {
+			return nil, err
+		}
+		copy(w.area, v.area)
+	}
+	return w, nil
+}
+
+// Shrink use to shrink vectors, sels must be guaranteed to be ordered
+func (v *Vector) Shrink(sels []int64) {
+	if v.class == FLAT {
+		switch v.typ.Oid {
```

Comment:
is it possible for us to have write this switch once?   I don't know, could be difficult for go.

Reply:
Yes, you can copy the []byte directly, but it may not be easy to understand if you write it this way. This might be a trade-off, I'm not sure using rawdata directly is a good choice.